# Project 2 - Customer Service Request Analysis

In [3]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

df = pd.read_csv("311_Service_Requests_from_2010_to_Present.csv")

/var/folders/jp/lmlcq7793rq6d18q105ksf480000gn/T/ipykernel_11572/3446402816.py:6: DtypeWarning: Columns (48,49) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("311_Service_Requests_from_2010_to_Present.csv")


## 1. Understand the Dataset
### 1.a Identify the shape of the dataset

In [5]:
df.shape

(364558, 53)

In [7]:
df.head()

Unique Key            Created Date             Closed Date Agency  \
0    32310363  12/31/2015 11:59:45 PM  01/01/2016 12:55:15 AM   NYPD   
1    32309934  12/31/2015 11:59:44 PM  01/01/2016 01:26:57 AM   NYPD   
2    32309159  12/31/2015 11:59:29 PM  01/01/2016 04:51:03 AM   NYPD   
3    32305098  12/31/2015 11:57:46 PM  01/01/2016 07:43:13 AM   NYPD   
4    32306529  12/31/2015 11:56:58 PM  01/01/2016 03:24:42 AM   NYPD   

                       Agency Name           Complaint Type  \
0  New York City Police Department  Noise - Street/Sidewalk   
1  New York City Police Department         Blocked Driveway   
2  New York City Police Department         Blocked Driveway   
3  New York City Police Department          Illegal Parking   
4  New York City Police Department          Illegal Parking   

                     Descriptor    Location Type  Incident Zip  \
0              Loud Music/Party  Street/Sidewalk       10034.0   
1                     No Access  Street/Sidewalk       11105.0   
2                     No Access  Street/Sidewalk       10458.0   
3  Commercial Overnight Parking  Street/Sidewalk       10461.0   
4              Blocked Sidewalk  Street/Sidewalk       11373.0   

        Incident Address  ... Bridge Highway Name Bridge Highway Direction  \
0    71 VERMILYEA AVENUE  ...                 NaN                      NaN   
1        27-07 23 AVENUE  ...                 NaN                      NaN   
2  2897 VALENTINE AVENUE  ...                 NaN                      NaN   
3    2940 BAISLEY AVENUE  ...                 NaN                      NaN   
4          87-14 57 ROAD  ...                 NaN                      NaN   

  Road Ramp Bridge Highway Segment Garage Lot Name Ferry Direction  \
0       NaN                    NaN             NaN             NaN   
1       NaN                    NaN             NaN             NaN   
2       NaN                    NaN             NaN             NaN   
3       NaN                    NaN             NaN             NaN   
4       NaN                    NaN             NaN             NaN   

  Ferry Terminal Name   Latitude  Longitude  \
0                 NaN  40.865682 -73.923501   
1                 NaN  40.775945 -73.915094   
2                 NaN  40.870325 -73.888525   
3                 NaN  40.835994 -73.828379   
4                 NaN  40.733060 -73.874170   

                                   Location  
0   (40.86568153633767, -73.92350095571744)  
1  (40.775945312321085, -73.91509393898605)  
2  (40.870324522111424, -73.88852464418646)  
3   (40.83599404683083, -73.82837939584206)  
4  (40.733059618956815, -73.87416975810375)  

[5 rows x 53 columns]

In [8]:
df.describe()

Unique Key   Incident Zip  X Coordinate (State Plane)  \
count  3.645580e+05  361560.000000                3.605280e+05   
mean   3.106595e+07   10858.496659                1.005043e+06   
std    7.331531e+05     578.263114                2.196362e+04   
min    2.960737e+07      83.000000                9.133570e+05   
25%    3.049938e+07   10314.000000                9.919460e+05   
50%    3.108795e+07   11209.000000                1.003470e+06   
75%    3.167433e+07   11238.000000                1.019134e+06   
max    3.231065e+07   11697.000000                1.067186e+06   

       Y Coordinate (State Plane)  School or Citywide Complaint  Vehicle Type  \
count               360528.000000                           0.0           0.0   
mean                203425.305782                           NaN           NaN   
std                  29842.192857                           NaN           NaN   
min                 121185.000000                           NaN           NaN   
25%                 182945.000000                           NaN           NaN   
50%                 201023.000000                           NaN           NaN   
75%                 222790.000000                           NaN           NaN   
max                 271876.000000                           NaN           NaN   

       Taxi Company Borough  Taxi Pick Up Location  Garage Lot Name  \
count                   0.0                    0.0              0.0   
mean                    NaN                    NaN              NaN   
std                     NaN                    NaN              NaN   
min                     NaN                    NaN              NaN   
25%                     NaN                    NaN              NaN   
50%                     NaN                    NaN              NaN   
75%                     NaN                    NaN              NaN   
max                     NaN                    NaN              NaN   

            Latitude      Longitude  
count  360528.000000  360528.000000  
mean       40.724980     -73.924946  
std         0.081907       0.079213  
min        40.499040     -74.254937  
25%        40.668742     -73.972253  
50%        40.718406     -73.930643  
75%        40.778166     -73.874098  
max        40.912869     -73.700715

#### The data set has 53 fields and 364,558 entries

### 1.b Identify the variables with null values

In [6]:
for col in df:
    null_vals = df[col].isna().sum(axis=0)
    if null_vals:
        print(col, " has ", null_vals, " null value(s)")

Closed Date  has  2381  null value(s)
Descriptor  has  6501  null value(s)
Location Type  has  133  null value(s)
Incident Zip  has  2998  null value(s)
Incident Address  has  51699  null value(s)
Street Name  has  51699  null value(s)
Cross Street 1  has  57188  null value(s)
Cross Street 2  has  57805  null value(s)
Intersection Street 1  has  313438  null value(s)
Intersection Street 2  has  314046  null value(s)
Address Type  has  3252  null value(s)
City  has  2997  null value(s)
Landmark  has  364183  null value(s)
Facility Type  has  2389  null value(s)
Due Date  has  3  null value(s)
Resolution Action Updated Date  has  2402  null value(s)
X Coordinate (State Plane)  has  4030  null value(s)
Y Coordinate (State Plane)  has  4030  null value(s)
School Region  has  1  null value(s)
School Code  has  1  null value(s)
School Zip  has  1  null value(s)
School or Citywide Complaint  has  364558  null value(s)
Vehicle Type  has  364558  null value(s)
Taxi Company Borough  has  364558 

## 2. Perform Basic Data Exporatory Analysis
### 2.a Utilize Missing Value Treatment

Perform the following steps:

1. Understand the dataset:

  a. Identify the shape of the dataset

  b. Identify variables with null values

2. Perform basic data exploratory analysis:

Utilize missing value treatment

Analyze the date column and remove the entries if it has an incorrect timeline

Draw a frequency plot for city-wise complaints

Draw scatter and hexbin plots for complaint concentration across Brooklyn

3. Find major types of complaints:

Plot a bar graph of count vs. complaint types

Find the top 10 types of complaints

Display the types of complaints in each city in a separate dataset

 

4. Visualize the major types of complaints in each city

5. Check if the average response time across various types of complaints

6. Identify significant variables by performing a statistical analysis using p-values and chi-square values (Optional)